### PR TITLE
Added (under testing) docs feature + refactored code (split into cogs)

### PR DESCRIPTION
**Pull request by infiniteregrets**
I am testing the ability to search the Manim documentation using the requests-html library. So far, it is definitely doable but will take me some time to refactor the code since the embed limit can exceed if the request result is over the 2048 char limit. 

I have refactored the code according to the concept mentioned here: https://discordpy.readthedocs.io/en/stable/ext/commands/extensions.html

This is probably the best practice, making the code more readable and allowing future extensions. 
If you guys like the way it is right now, do let me know! I have some more ideas to make it more readable!

I was unable to test the !manimate command, it always errored out on my machine. I am sure someone can test it out. Do let me know if it works and if possible please make appropriate changes to make it work in case I missed anything.

Thank you, everyone! Let me know what you all think (:

### CODE REVIEW

**Comment by behackl** [comment]
This looks very nice, thanks for your efforts of making the bot more maintainable!

I didn't review the changes in detail, but I'm pretty sure that the direction is good. If you would like me (or someone else) to review this in more detail, just let us know!

**Comment by infiniteregrets** [comment]
hey! it would be awesome if someone could test them out and let me know if they would like any more changes?

File: cogs/mhelp.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+!manimate -s
+\`\`\`py
+def construct(self):
+    self.play(ReplacementTransform(Square(), Circle()))
+\`\`\`
+```
+""")
+
+def setup(bot):
+    bot.add_cog(Help(bot))
```
Comment (Aathish04) [inline cogs/mhelp.py:31]:
```suggestion
    bot.add_cog(Help(bot))

```
PEP8 recommends that you end every file with a newline.

File: config.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+PREFIX = '!'
+TOKEN = ''
+#Todo -- update class dictionary
```
Comment (Aathish04) [inline config.py:3]:
```suggestion
#Todo -- update class dictionary
```
Remove a trailing space. On second thought, could you please remove the trailing spaces from every file? PEP8 recommends that all trailing whitespaces should be removed.

Usually, your editor/IDE will have an option for this.

File: cogs/mdoc.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            self.title = f'`Results for: {arg}`'
+
+        embed = discord.Embed(title = self.title, 
+                            description = '\n'.join(self.res),
+                            color = 0xe8e3e3)
+
+        return await ctx.reply(embed = embed, mention_author=False)                            
+
+def setup(bot):
+    bot.add_cog(Mdoc(bot))
```
Comment (Aathish04) [inline cogs/mdoc.py:53]:
```suggestion
    bot.add_cog(Mdoc(bot))

```

PEP8 recommends that you end every file with a newline.

**Review by Aathish04 — DISMISSED** [review]
Some formatting changes.

Could you please remove the trailing spaces from every file? PEP8 recommends that all trailing whitespaces should be removed.

Usually, your editor/IDE will have an option for this.

**Review by infiniteregrets — COMMENTED** [review]
Thanks for the suggestion! Maybe using something like flake8 would be great?

**Comment by Darylgolden** [comment]
Is it worth setting up the precommit hook here too?

**Comment by infiniteregrets** [comment]
> Is it worth setting up the precommit hook here too?
 
I am not sure about that, but whatever is a part of your workflow i'd be happy to follow 

Did anyone get a chance to test out the changes?

**Comment by Darylgolden** [comment]
When I try to run it, I get
```mdoc.py couldn't be loaded
__pycache__ couldn't be loaded```
Any idea what could be the issue?

**Comment by Darylgolden** [comment]
But the `!manimate` command does work properly for me.

**Comment by Darylgolden** [comment]
As mentioned in my Discord test server, I'll wait for more changes before reviewing it again.

**Comment by infiniteregrets** [comment]
> As mentioned in my Discord test server, I'll wait for more changes before reviewing it again.

Sure!

**Comment by infiniteregrets** [comment]
I have added the pagination. A bit buggy right now, but will be fixed soon

**Comment by behackl** [comment]
What is the status with this PR? Is there anything we can do to help move things along? Given how much this improves the current code structure, I'd be certainly okay with merging a non-polished version and delay further improvements to follow-up PRs. :-)

**Comment by infiniteregrets** [comment]
Hey, I haven't been able to find time to fix up some bugs but I am sure I can get all this done by the weekend. 
There are a few things I'd like to be added/fixed which I need to get done:
1. Add the new docstring command as a cog. (if the person who implemented this could add it then it would be great)
2. Polygon doc command fails to return a link
3. Fix reactions on mdoc and manimate 
4. Clear the cooldown command cache (There is a new cooldown feature now that doesn't allow users to abuse the bot, for example the manimate command)
5. Most importantly, add log configuration to all the modules and I want to try this feature I had in mind that allows you to send log files via email which I think could be helpful to @behackl while he sets up the bot on his machine.

If I am stuck anywhere, then I will probably ping you guys. Thanks (:

**Comment by chirag-droid** [comment]
Hey, I found the possible reason why reactions are not working, see my above review comments. That should hopefully fix it. Thanks

**Comment by chirag-droid** [comment]
> Hey, I found the possible reason why reactions are not working, see my above review comments. That should hopefully fix it. Thanks

Sorry didn't know my review wasn't submitted I was using GitHub Android client and it didn't show it

File: cogs/manimate.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                    return reply_args
+
+        async def react_and_wait(reply):
+            await reply.add_reaction("\U0001F5D1") # Trashcan emoji
+
+            def check(reaction, user):
+                return str(reaction.emoji) == '\U0001F5D1' and user == ctx.author
+
+            try:
+                reaction, user = await bot.wait_for('reaction_add', check=check, timeout=60.0)
```
[the remark this answers is not in the document]
Comment (infiniteregrets) [reply cogs/manimate.py:124]: hey, good catch! when I was porting it I probably forgot to fix it. TYVM

File: DiscordManimator.py [hunk 5]
```diff
@@ -1,210 +1,46 @@
...
-                            )
+class Manim(commands.Bot):
+    def __init__(self):                     
+        manim_intents = {
+                        'guilds': True, 
+                        'members': True, 
+                        'messages': True, 
+                        'reactions': True, 
+                        'presences': True
                         }
```
[the remark this answers is not in the document]
Comment (infiniteregrets) [reply DiscordManimator.py:18]: I just selected a few I thought we might need in the future. Will fix that

**Comment by naveen521kk** [comment]
@infiniteregrets if not already can you format the code with [black](https://black.readthedocs.io)+[isort](https://pycqa.github.io/isort/)?

**Comment by Darylgolden** [comment]
I've asked @infiniteregrets on the status of this PR and they said that they were busy with studies. How about merging this to a separate branch for now so others can work on it?

**Comment by chirag-droid** [comment]
Please merge this to a separate branch its been so long.

**Comment by Darylgolden** [comment]
@chirag-droid I've merged it.

**Pull request merged**